Rename the core provided fields to localgov_ prefix

andybroomfield opened the pull request:
Fix #35

This will rename the config fields in localgov_core (not the media or other
modules).
This also provides an update hook to rename the database tables and replace 
the config.

WARNING! Applying this will require an update in localgov_services and other
places that use these field.

andybroomfield committed:
Rename the field storage config to localgov_ prefix

Ref #35

Note: This is only the storage, and the actual field usage is in indivdual
modules (mostly localgov_services). These are going to have to be updated in
tandem.

andybroomfield committed:
Update function to rename the core fields

Ref #35

Warning! Applying this will mean that the fields usage in Localgov Services
will need to be updated.

andybroomfield commented:
@stephen-cox @finnlewis This is a very basic implementation to rename the fields in core.
There is a problem with this, and fields in https://github.com/localgovdrupal/localgov_services/issues/75 are a lot more complicated to update and the config will need to be rewritten. 
https://www.drupal.org/project/config_sync and migration may be a better method for handling this.

stephen-cox reviewed (commented):
Running this with the localgov_demo module enable that includes data for these fields, it appears that the field is renamed and the data moved, but the field loses it's association with the given content type. That is, the database tables exits and Drupal can see the field, but it's not listed as  as one of the managed fields of the localgov_services_landing content type.

@andybroomfield  If we could get this way to work it might be simpler than trying to migrate data with partial configuration imports. Are your concerns with this based on the number of potentially risky database changes needed here?

andybroomfield commented:
@stephen-cox This is going to need someone with a lot more domain knowledge of where these fields are used to make the change. At least services, but potentially any node that uses the old field has to be upgraded. I'm not sure what the best statergey for that would be, to find and load the node config or to  update the bundle in the database?
This is also cross module as this module just provides the field storage, but the usuage is in localgov_services (and other places).

stephen-cox commented:
I have had another look at this and hoped to get something working, but have run out of time. What I think we need to do when renaming fields is after you've updated the field storage config and moved the date, as in this pull request, is to load all the dependant config entities, rename the field.field.node.* ones and update the core.entity_form_display.node.* and core.entity_view_display.node.* ones for all bundles using the field.

You can get the dependant config entities with:

```php
$config_manager = Drupal::service('config.manager');
$field_storage_name = $field_storage->getConfigDependencyName();
$dependents = $config_manager->findConfigEntityDependentsAsEntities('config', [$field_storage_name]);
foreach ($dependents as $dependent) {
  /** @var \Drupal\Core\Config\Entity\ConfigEntityInterface $dependent */
}
```

I'll take another look at this when I have some time.

Pinging @ekes as he'll might have some thoughts on this :wink:

andybroomfield commented:
Taking another look at this as we may need to start incorporating modules directly into BHCC site.

andybroomfield committed:
WIP: Load the dependant config and apply to renamed fields

- Load up each depended config
- Create a new field for the replacement field
- Add to the manage form and manage display screen

andybroomfield commented:
I made some progress with @stephen-cox suggestion. 
This does now copy the field config over, but does not replace them into the correct groups.

Pre Update:
<img width="1218" alt="Screenshot 2021-01-07 at 11 19 03 am" src="https://user-images.githubusercontent.com/1467480/103887143-8175b680-50da-11eb-8e01-1094460f7f35.png">

Post Update: Empty group:
<img width="1215" alt="Screenshot 2021-01-07 at 11 19 22 am" src="https://user-images.githubusercontent.com/1467480/103887152-85093d80-50da-11eb-8681-641fee4cfd5e.png">

Fields dumped at end:
<img width="1228" alt="Screenshot 2021-01-07 at 11 19 28 am" src="https://user-images.githubusercontent.com/1467480/103887161-889cc480-50da-11eb-9ddf-ab6d0905760c.png">

andybroomfield committed:
WIP: Fix renamed table data values

Fix issue with missing table field (needs localgov_FIELDNAME_value).
Fix base table missing by removing them after all the renaming done.

andybroomfield commented:
I've managed to get this to get a super very rough version of this to work.
Fields machine name is renamed and the data moved accross and the fields are present on the manage form / display screen.
The templates need to be updated to account for the new field.
There is an issue with correctly placing the replacement fields into the correct groups on the manage form display screen from the update.

It would be good if someone could take a look over my rough code to provide some future direction on how to refactor this.

<img width="676" alt="Screenshot 2021-01-07 at 2 35 43 pm" src="https://user-images.githubusercontent.com/1467480/103905112-19cd6480-50f6-11eb-99bd-94b8b470e6c1.png">

iam-jam commented:
Hi @andybroomfield 

I've looked into the issue with field groups, and while I've not tested it locally, I may be able to offer some pointers. 

When exporting a content type with field groups the config entity includes something similar to the following:

```
third_party_settings:
  field_group:
    group_test:
      children:
        - title
      parent_name: ''
      weight: 0
      format_type: fieldset
      region: content
      format_settings:
        id: ''
        classes: ''
        description: ''
        required_fields: true
      label: Test
```
As you can see, this includes a `children` key which will need to be altered to use the new field names. `EntityFormDisplay` provides a `getThirdPartySetting` and a `setThirdPartySetting` so you should be able to do something like:

```
$fieldGroupConfig = $dependent->getThirdPartySetting('field_group', 'group_test'); // Where group_test is the machine name of the field_group
foreach($fieldGroupConfig['children'] $key as $name) {
   if ($name == 'old_name') {
      $fieldGroupConfig['children'][$key] = 'new_name';
  }
}
$dependent->setThirdPartySetting('field_group', 'group_test', $fieldGroupConfig);
```

If you don't know which field groups you're going to acting on then obviously this is going to require searching for field group settings in the dependent configs first.

andybroomfield commented:
Thanks @iam-jam I'm going to try that.

andybroomfield committed:
Preserve field groups when updating field machine names

Loop through each field group and if the field is present,
update name in group

andybroomfield commented:
Although a draft, if anyone wants to review current progress (@iam-jam @stephen-cox) that would be appreciated.

I have it now updating the field machine names and updating the node field groups so its retained.
Steps to test
- Use the master / default install to install localgov and enable the localgov_demo module
- Verifiy the old field machine names on the localgov_services_landing_page for field_email_address, field_facebook, field_twitter, field_phone
- Switch to this branch  fix/35-rename-fields
- run database updates (lando drush updb)
- Verify the field machine names on localgov_services_landing_page are replaced with localgov_email_address, localgov_facebook, localgov_twitter, localgov_phone
- goto /node/1/edit and check that the fields are present in the right groups (social media and contact) with values preservied.

This won't display on front end as thats coded in the template for the services landing page.
To use, will also need amends to localgov_services module.

I've been thinking we need to turn this into a helper utility function to rename any field, and then call that in the update hooks. Suggestions on best practice and how to refactor this welcome.

iam-jam commented:
Hi @andybroomfield 

I've taken another look and a couple of things have occurred to me:

1. Before executing any SQL, such as `INSERT INTO $new_table SELECT * FROM $old_table`, I might be a good idea to check that the fields we're trying to rename exist - perhaps this edge case? - this could be as simple as checking the return value from `FieldStorageConfig::loadByName`.
2. Exception handling - relating to the above and probably other API calls, we might want to throw a an `UpdateException` (https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Utility%21UpdateException.php/class/UpdateException/8.2.x) in places.

Hope this helps @andybroomfield, let me know if you want to talk anything over.

andybroomfield commented:
Thanks @iam-jam 
Yes I think we need to check and abort if a renamed field already exists, alothough we have to delete the table first as installing the new field config doe set up the table, and we need to clone an existing one to copy into it.
What I'm going to do is make this a generic function that other update hooks can call into, and it can abort out of that the update hooks can call an update exception.

Adnan-cds commented:
Hi Andy,
Sorry, can't offer any help on this.  Knee deep in migration at the moment :(  But happy to try out the field renaming update function when you feel it is ready.

I am assuming this change will take care of field renaming in Drupal config, config yamls and PHP files (e.g. tests).  I am prepared to take the hit on the theme templates.  Hopefully grep/find/sed would make things slightly easier.

Thanks for working on it :)

andybroomfield commented:
Thanks @Adnan-cds Whilst this PR renames and updates the config, as the field usage is spread accross the localgov_services modules a co-ordinated update will also need to be built and run to update the code usage, and localgov_theme will also need to update any template usages. Hopefully should be a search and replace. 
@stephen-cox was working on a test branch to update the usuage of these fields in localgov_services.
https://github.com/localgovdrupal/localgov_services/compare/fix/75-rename-fields

andybroomfield commented:
If the main concern is twig templates, I did a quick experiment and the following shim will work.
Twig templates only.
````php
function localgov_services_landing_preprocess_node(&$vars) {
  $shim_field_names = ['email_address', 'facebook', 'phone', 'twitter'];
  foreach ($shim_field_names as $name) {
    $old_field_name = 'field_' . $name;
    $new_field_name = 'localgov_' . $name;
    if (empty($vars['content'][$old_field_name]) && !empty($vars['content'][$new_field_name])) {
      $vars['content'][$old_field_name] = $vars['content'][$new_field_name];
    }
  }
}
````

andybroomfield committed:
Initial move the field rename logic into a seperate class for reuse

andybroomfield committed:
Check the correct DB column names to update field data

andybroomfield committed:
Update the docblock on localgov_core update hook

Descripe the fields that will be updated

andybroomfield commented:
@danchamp This is the PR for updating and renaming fields if you would like to test.
[See comment above on steps to test](https://github.com/localgovdrupal/localgov_core/pull/37#issuecomment-760985207). 

I'm also preparing a companion PR in Localgov Services to rename the fields their https://github.com/localgovdrupal/localgov_services/pull/91

danchamp commented:
@andybroomfield I've just tested this on the London Councils dev site, which has a fully implemented services section.

From initial review the updates have been applied with no issues. The update itself completed successfully, with all database tables and fields renamed as expected. All existing content are functionally as-was on the front-end, and editing forms operate without issue. All code tests pass locally.

This is the only anomaly I see in services now, but it wasn't introduced as part of this work:

https://github.com/localgovdrupal/localgov_services/blob/fix/75-rename-fields/modules/localgov_services_landing/config/install/field.storage.node.field_localgov_services_landing.yml

From what I can make out field_localgov_services_landing is orphaned and unused, and could be removed (via an update hook).

Great work on this.

andybroomfield commented:
Thanks @danchamp good to hear it is working.

Does this also include the PR localgovdrupal/localgov_services#91 and the theme fixes in localgovdrupal/localgov_theme#153 ?

I had wondered about that odd field_localgov_services_landing. Seems to be an old field left over during the initial conversion of this module. As long as nobody is using that field, it should be ok in the localgov_services PR to include the removal of that field.

I'll wait for feedback from other sites on what issues arise from this PR. The next steps here are
- Include some more error checking in case the renamed fields (of localgov_* variety) already exist.
- An automated test for the field rename helper utility to verify it works in the test suite.

danchamp commented:
> Does this also include the PR [localgovdrupal/localgov_services#91](https://github.com/localgovdrupal/localgov_services/pull/91) and the theme fixes in [localgovdrupal/localgov_theme#153](https://github.com/localgovdrupal/localgov_theme/pull/153) ?

Yes, I ran the updates and tested with those branches checked out.

> I had wondered about that odd field_localgov_services_landing. Seems to be an old field left over during the initial conversion of this module. As long as nobody is using that field, it should be ok in the localgov_services PR to include the removal of that field.

Its database tables are empty in all of my installs, and I can't find any reference to it in code.

> I'll wait for feedback from other sites on what issues arise from this PR. The next steps here are
> 
>     * Include some more error checking in case the renamed fields (of localgov_* variety) already exist.
>     * An automated test for the field rename helper utility to verify it works in the test suite.

I don't have time to help with the specifics just now I'm afraid, but happy to review and test where I can.

paulpopus commented:
Hi @andybroomfield sorry to get this back to you late.

I've ran these updates from [localgov_services PR](https://github.com/localgovdrupal/localgov_services/pull/91) and this module together against a copy of the Lambeth site and I ran into a couple of issues.

```bash
>  [error]  Call to a member function toArray() on null 
>  [error]  Update failed: localgov_core_update_8001 
>  [notice] Update started: localgov_services_landing_update_8001
>  [error]  Call to a member function toArray() on null 
>  [error]  Update failed: localgov_services_landing_update_8001 
```

Seems to be related to the `FieldRenameHelper.php` on l45 and l84.

I haven't done more debugging to figure out why, but happy to if you need more help with this. The other update hooks work without any issues.

andybroomfield commented:
Thanks @paulpopus 
Line 45 
````php
$new_field_storage = $field_storage->toArray();
````
and line 84
````php
$new_field = $dependent->toArray();
````
That suggests it would be trying to fetch a field that does not exist.
Do you have the following fields? 
>     'address_first_line',
>     'address',
>     'contact_us_online',
>     'destinations',
>     'hearing_difficulties_phone',
>     'link_to_map',
>     'opening_hours',
>     'other_team_contacts',
>     'popular_topics',

If any are missing, I think that is the failure, and the fix will be to check if the field exists first.

paulpopus commented:
@andybroomfield 
Right, so some of those fields are missing with only these remaining on the landing page for example
```
'destinations',
'popular_topics',
```
Some of the other (I assume reused) fields did migrate on the landing page though but it sounds like you're definitely right about needing to check if the field exists.

andybroomfield committed:
Fix field rename error when field is missing

Check if field is present before renaming

andybroomfield commented:
@paulpopus Pushed a fix that should resolve this, let me know if you have a chance to test again

paulpopus commented:
Thanks @andybroomfield they all migrate successfully now!

andybroomfield commented:
Thanks @paulpopus Any issues with the Lambeth site itself, or the scale of the work needed to get the site working after applying this update?

paulpopus commented:
@andybroomfield from what I've experienced so far, there isn't anything else that needs to be done except to check in the theme that these renamed fields are not preprocessed or included by machine name in the templates if you have a custom theme. Which for lambeth do exist in the .theme file.

andybroomfield commented:
@paulpopus ok, thanks. Part of what this excerisce is finding out what impact this change will have to councils who have already built something with the distribution. Yes any custom module or theme code that calls the machine name needs to be updated so it would be good to know if thats going to be a major and common thing.

andybroomfield committed:
Add a test for the field rename helper

Adds a kernel test that generates a node type with a field, renames it and
checks the fields in the node type has been renamed

andybroomfield committed:
Fix to make the field rename database queries work with table prefix

Adnan-cds commented on src/FieldRenameHelper.php:
```diff
@@ -0,0 +1,134 @@
+    // @TODO find out how to get the correct table names.
+    $old_table = $entity_type . '__' . $old_field_name;
+    $new_table = $entity_type . '__' . $new_field_name;
+    $revision_old_table = $entity_type . '_revision__' . $old_field_name;
+    $revision_new_table = $entity_type . '_revision__' . $new_field_name;
```
This is only relevant when ```$field_storage->isRevisionalbe()``` applies.  But probably won't do us much harm, so can be left intact :)

andybroomfield replied:
For the moment these are nodes, however would be useful in future.

Adnan-cds reviewed (commented):
Herculean effort Andy :100: :+1: 

I have a feeling we should replace as much of the direct SQL queries with their API equivalents.  I will see if I can find the right API calls.  I will get back to you on this whatever the outcome.

We are not too far :)

Adnan-cds commented on src/FieldRenameHelper.php:
```diff
@@ -0,0 +1,134 @@
+    $new_field_name = substr($new_field_name, 0, 32);
+
+    // Field variables.
+    // @TODO find out how to get the correct table names.
```
Perhaps something like this?:
```
$field_storage_key_name = $entity_type . '.field_schema_data.' . $old_field_name;
$key_value_storage = Drupal::keyValue('entity.storage_schema.sql');
$storage_schema = $key_value_storage->get($field_storage_key_name);
$existing_table_names = array_keys($storage_schema);
```

I am trying to find out if we need to update relevant items within the *entity.storage_schema.sql* key's value as part of field renaming.

Adnan-cds replied:
Hi Andy,
I have added some code to copy the fields based on the above.  Hope you don't mind.

I still have some more manual data validation to do.

andybroomfield replied:
This looks good now so resolving this.

andybroomfield commented on src/FieldRenameHelper.php:
```diff
@@ -0,0 +1,134 @@
+    foreach($field_columns as $column) {
+      $old_table_value_column = $old_field_name . '_' . $column;
+      $new_table_value_column = $new_field_name . '_' . $column;
+      $sql[] = "ALTER TABLE $prefix$new_table CHANGE $old_table_value_column $new_table_value_column varchar(255)";
+      $sql[] = "ALTER TABLE $prefix$revision_new_table CHANGE $old_table_value_column $new_table_value_column varchar(255)";
```
@Adnan-cds I'd like to find a way of doing this that is a lot safer. So far I didn't have luck trying to find something that would would work outside raw SQL.

To understand what this is doing: Creating the new field config creates the new table, but we need an identical table to do the copy and then rename it, so we delete it first, create a copy like the old one and copy all the data over, then we alter the column for the value field.
The varchar thing is a bit strange. In MySQL 8 Its not needed and the column is simply renamed, however mysql 5.7 requires at least a value for the statement to work. What we need here is a way to get the original values for this field and apply it here (or find a better way of doing the copy).

Adnan-cds replied:
Got it.  I am trying to come up with some code.  Hopefully tomorrow.

andybroomfield replied:
Actully, I did a little checking on this, it turns out (at least in the Lando install) that the only 2 sql lines needed are
````php
$sql[] = "INSERT INTO $prefix$new_table SELECT * FROM $prefix$old_table";
````
````php
$sql[] = "INSERT INTO $prefix$revision_new_table SELECT * FROM $prefix$revision_old_table";
````
And all the other tables stay intact. I wonder if I do have an exotic setup somewhere that required it that way?
This means these lines, and the delete and recreate the table SQL can be dropped.

Adnan-cds replied:
> the only 2 sql lines needed are

Very true :)

andybroomfield replied:
@Adnan-cds I've updated this with the superfluous sql statements removed. Let me know if that still works.

andybroomfield commented on src/FieldRenameHelper.php:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    foreach ($sql as $indv_sql) {
+      $database->query($indv_sql);
```
I'm also thinking it would be safer to execute each statement indivudually instead of doing it all together in a for each.

andybroomfield replied:
I'm seeing the logging code in your changes so looks good here.

Adnan-cds replied:
Thanks Andy.  I have made some more changes to satisfy the code sniffer.  I don't think I have any more code changes to offer.  This PR should now pass the tests if the related PR from localgov_services were available.  In any case, I will keep checking the database manually to ensure no data or config is lost.  I will keep you posted.

andybroomfield committed:
Remove superfluous sql statements as just and insert all needed

- Remove the drop and recreate table
- Remove the alter column name
This relies on fieldconfig::create to create the correct table structure

Adnan-cds committed:
Field renaming: Use schema for data copying.

When renaming fields, we need to copy data between the tables of the old and new
field.  Drupal's database schema can tell us the tables associated with a field.
This helps to determine which table's data should be copied to where.

Adnan-cds committed:
Temporary change to test runner.

To be reverted once #51 is merged.

Adnan-cds committed:
Field renaming: Coding standard changes.

More comments suggesting 'drush config:export' be run after the update hook is
complete.  This is so that 'drush deploy' doesn't revert the changes introduced
by this update.

andybroomfield commented:
For the test, its also going to require pulling in the services PR localgovdrupal/localgov_services#91
@Adnan-cds

andybroomfield committed:
Initial add test for field group in entity display

andybroomfield committed:
Set up field groups in form and view display and test for renamed field

Use set third party settings to add the field group and check its
presence with get third party settings

andybroomfield committed:
Add test for field config (weight, label) in each display

- Check the weight is preserved in form display
- Check the weigth and label are preserved in view display

andybroomfield commented:
Next thing is to add tests for checking that the field config and field groups are preserved during the rename. 
I've got some code written to do this, though it relies on field_group being installed, which isn't a dependancy in localgov_core. Unless I can figure out another way would it ok to add that as a dependancy?

Adnan-cds commented:
> which isn't a dependancy in localgov_core.

That must be an [oversight](https://github.com/localgovdrupal/localgov_core/blob/783cdcae8a64f6f54f9eb9c97eb480a3373d9d76/composer.json#L9).  Feel free to add it to the [info file](https://github.com/localgovdrupal/localgov_core/blob/783cdcae8a64f6f54f9eb9c97eb480a3373d9d76/localgov_core.info.yml#L7).

andybroomfield committed:
Add field group as dependancy

andybroomfield committed:
Code standard fix

Adnan-cds commented:
Things are looking good Andy.  I have tried renaming *field_page_components* from the localgov_page_components module as a test and even that worked fine.  I haven't lost any data.  So final :+1: from me :)

While checking various parts of Drupal **after** running the update hook, I noticed that ```\Drupal::keyValue('entity.storage_schema.sql')->get('node.field_schema_data.field_phone)``` still retains records for the renamed field_phone and other fields.  Further testing led me to [field_purge_batch()](https://api.drupal.org/api/drupal/core%21modules%21field%21field.purge.inc/function/field_purge_batch/9.1.x) from the *field* module.  This funtion revealed that all residues of a deleted field are only removed during the next config **import**.  So I did a 'drush config:export', refreshed the database of my dev site and then a ```drush deploy```.  The fields got renamed as expected and all mentions of the old fields were gone.  We don't have to explicitly call field_purge_batch() in FieldRenameHelper::renameField(), but it was an interesting read, so I thought I will mention it here.

Once again, thanks for bring us this far :)

willguv commented:
@andybroomfield @Adnan-cds I met with @ReeceMarsland and Alasdair to talk about the impact on Westminster just now. Reece is looking, but he doesn't think it's too onerous. He'll post any questions or thoughts. 

Let's pick this up at the next PG - think we're almost good to go.

andybroomfield committed:
Add message to export configuration once update is completed

andybroomfield committed:
Real delete the test.yml file
